### PR TITLE
added seedset function to prototype selection algorithms

### PR DESCRIPTION
**Pull request by qiyunzhu**
Hi @sjanssen2 I added the "seedset" parameter to three of the remaining four prototype selection functions. Please read with critics, as I am not 100% sure if the code I wrote does what it's supposed to. I couldn't figure out how to implement it for the "proto class" function. Maybe you would like to do it? In addition, I noticed that the "pMedian" function produces quite different prototypes with notably smaller sum of distances. Is that normal? Furthermore, I added a "validate parameter" function and let all functions call it.

### CODE REVIEW

**Comment by sjanssen2** [comment]
@qiyunzhu that is impressive work. Thanks so much. I especially like your porting of the input arg validation into its own function.
However, it is hard to access to what degree seedsets influence runtime. Therefore, I would like to first come up with some benchmarks for my original functions. Then re-run the benchmark with seedsets and check how much overhead seedsets impose. Thus, I'd like to block your PR for some time, just to give me the necessary dev time for the benchmarking. Is that OK?

**Review by sjanssen2 — CHANGES_REQUESTED** [review]
give me time :-)

**Comment by sjanssen2** [comment]
With my benchmarking notebook I could now access that runtime without any seedsets is not compromised by your changes. That's awesome!

File: genomesubsampler/prototypeSelection.py [hunk 1]
```diff
@@ -41,6 +41,56 @@
...
+    ----------
+    dm: skbio.stats.distance.DistanceMatrix
+        Pairwise distances for all elements in the full set S.
+    num_prototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+    seedset: iterable of str
+        A set of element IDs that are preferably selected for the resulting
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: 'preferably' does not seem to be the right wording, because we ensure that those elements go in first.

File: genomesubsampler/prototypeSelection.py [hunk 2]
```diff
@@ -41,6 +41,56 @@
...
+                         "prototype is useless.")
+    if num_prototypes >= dm.shape[0]:
+        raise ValueError("'num_prototypes' must be smaller than the number of "
+                         "elements in the distance matrix, otherwise no "
+                         "reduction is necessary.")
+    if len(seedset) > 0:
+        seeds = set(seedset)
+        if len(seeds) < len(seedset):
+            raise ValueError("There are duplicated IDs in 'seedset'.")
+        if not seeds < set(dm.ids):
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: are you comparing set sizes here? I'd prefer to be more specific for readability and write:
`if len(seeds) < len(set(dm.ids)):`

File: genomesubsampler/prototypeSelection.py [hunk 3]
```diff
@@ -41,6 +41,56 @@
...
+            raise ValueError("'seedset' is not a subset of the element IDs in "
+                             "the distance matrix.")
+        if len(seeds) >= num_prototypes:
+            raise ValueError("Size of 'seedset' must be smaller than the "
+                             "number of prototypes to select, otherwise no "
+                             "selection is necessary.")
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: wording: `otherwise we cannot guarantee that all seed element are selected as prototypes. Would require a secondary objective.`

File: genomesubsampler/prototypeSelection.py [hunk 4]
```diff
@@ -114,30 +169,27 @@ def prototype_selection_exhaustive(dm, num_prototypes,
...
+    ids = set(dm.ids) - set(seedset)
+    num_prototypes = num_prototypes - len(seedset)
+    num_combinations = sp.special.binom(len(ids), num_prototypes)
     if num_combinations >= max_combinations_to_test:
         raise RuntimeError(("Cowardly refuse to test %i combinations. Use a "
                             "heuristic implementation for instances with more "
                             "than %i combinations instead!")
                            % (num_combinations, max_combinations_to_test))
 
+    seedset = tuple(seedset)
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: why is this necessary?

File: genomesubsampler/prototypeSelection.py [hunk 5]
```diff
@@ -114,30 +169,27 @@ def prototype_selection_exhaustive(dm, num_prototypes,
...
     max_dist, max_set = -1 * np.infty, None
-    for s in set(combinations(dm.ids, num_prototypes)):
-        d = distance_sum(s, dm)
+    for s in set(combinations(ids, num_prototypes)):
+        d = distance_sum(s + seedset, dm)
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py:185]: we should be able to save some compute time by ignoring seedset for distance_sum computation, because it is static within the loop.

File: genomesubsampler/prototypeSelection.py [hunk 6]
```diff
@@ -180,28 +238,32 @@ def prototype_selection_constructive_maxdist(dm, num_prototypes):
...
-                          "reduction is necessary."))
+    _validate_parameters(dm, num_prototypes, seedset)
 
     # initially mark all elements as uncovered, i.e. as not being a prototype
     uncovered = np.asarray([np.True_] * dm.shape[0])
+    res_set, num_found_prototypes = [], 0
+
+    if len(seedset):
+        # mark elements in the seedset as found
+        seedset = set(seedset)
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: this line should never have an effect, because of parameter validation

File: genomesubsampler/prototypeSelection.py [hunk 7]
```diff
@@ -180,28 +238,32 @@ def prototype_selection_constructive_maxdist(dm, num_prototypes):
...
-        raise ValueError(("'num_prototypes' must be smaller than the number of"
-                          " elements in the distance matrix, otherwise no "
-                          "reduction is necessary."))
+    _validate_parameters(dm, num_prototypes, seedset)
 
     # initially mark all elements as uncovered, i.e. as not being a prototype
     uncovered = np.asarray([np.True_] * dm.shape[0])
+    res_set, num_found_prototypes = [], 0
+
+    if len(seedset):
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: I'd rewrite the if block as:
With that we don't have to enumerate all maybe 80k ids of the distance matrix. (I don't know how efficient .index() is). And we can avoid the element by element append for the seedset

```if len(seedset):
    for _id in seedset:
        uncovered[(dm.ids).index(_id)] = np.False_
    res_set = list(seedset)```

File: genomesubsampler/prototypeSelection.py [hunk 8]
```diff
@@ -180,28 +238,32 @@ def prototype_selection_constructive_maxdist(dm, num_prototypes):
     def prototype_selection_constructive_maxdist(dm: DistanceMatrix,
     num_prototypes: int) -> List[str]:
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: we should add `, seedset: List[str]`

File: genomesubsampler/prototypeSelection.py [hunk 9]
```diff
@@ -329,13 +391,7 @@ def prototype_selection_constructive_protoclass(dm, num_prototypes, steps=100):
...
     num_prototypes: int, steps=100: int) -> List[str]:
     '''
-    if num_prototypes < 2:
-        raise ValueError(("'num_prototypes' must be >= 2, since a single "
-                          "prototype is useless."))
-    if num_prototypes >= len(dm.ids):
-        raise ValueError(("'num_prototypes' must be smaller than the number of"
-                          " elements in the distance matrix, otherwise no "
-                          "reduction is necessary."))
+    _validate_parameters(dm, num_prototypes)
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py:391]: no seedset for this function?? Can you open an issue for that such that I can work on it later on?

File: genomesubsampler/prototypeSelection.py [hunk 10]
```diff
@@ -426,20 +488,19 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
...
 
     # start with an empty list of prototypes
     prototypes = []
 
-    # add the one element whose distance is smallest to all other elements as
-    # the first prototype.
-    prototypes.append(np.argmin(dm.data.sum(axis=1)))
+    if len(seedset):
+        # pre-populate the prototype list with seeds
+        seedset = set(seedset)
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: again: not necessary due to validation

File: genomesubsampler/prototypeSelection.py [hunk 11]
```diff
@@ -426,20 +488,19 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
...
+        prototypes = [idx for idx, id in enumerate(dm.ids) if id in seedset]
```
Comment (sjanssen2) [inline genomesubsampler/prototypeSelection.py]: I'd like to avoid enumeration of all IDs, therefore I suggest:
```prototypes = [ (dm.ids).index(_id) for _id in seedset ]```

**Review by sjanssen2 — COMMENTED** [review]
Thanks for your good work. Here are some suggestions for further improvement.

File: genomesubsampler/prototypeSelection.py (hunk 2 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: No. This is to test whether set A is a subset of set B.

File: genomesubsampler/prototypeSelection.py (hunk 1 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: I think I wasn't the one who wrote "preferrably"... Could you suggest a better word? Maybe "guaranteed" or "pre"?

File: genomesubsampler/prototypeSelection.py (hunk 3 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: I think this error message doesn't have to be that complicated. Saying "Size of 'seedset' must be smaller than the number of prototypes to select." should be enough. The user just needs to know the rule, but doesn't have to understand the mechanism, from a simple error message.

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py:185]: I agree. Is there a good way to do it?

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: It is to improve performance.

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: I have changed the acceptable datatype of `seedset` into any iterable. So this is necessary.

File: genomesubsampler/prototypeSelection.py (hunk 9 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py:391]: I could not think of an easy way to do seedsetting for this function. Yes I will open an issue once the current PR is being merged.

File: genomesubsampler/prototypeSelection.py (hunk 10 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: See above.

File: genomesubsampler/prototypeSelection.py (hunk 11 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Cool! I didn't know this trick.

**Review by qiyunzhu — COMMENTED** [review]
Thanks for reviewing! I made responses to most (not all) of your comments.

File: genomesubsampler/prototypeSelection.py (hunk 1 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: Something like:
`A set of element IDs that is guaranteed to be selected as first prototypes prior to applying the prototype selection algorithm. Remaining prototypes are then recruited with the prototype selection algorithm.`

File: genomesubsampler/prototypeSelection.py (hunk 2 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: oh, magic. I didn't understand this from the code because I was not aware of the operator < definition. Should we add a comment line, because others might be confused as well or is it just me?

File: genomesubsampler/prototypeSelection.py (hunk 3 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: I agree! And therefore would remove the 'otherwise no selection is necessary' part, because this explanation is not only too short but also a bit wrong

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: I don't see why an additional conversion from list to tuple should improve performance?

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py:185]: just remove `+ seedset` in the `distance_sum()` call

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: why? Shouldn't matter if we iterate over a list (which is guaranteed to have no duplicate elements) or a set.

File: genomesubsampler/prototypeSelection.py (hunk 1 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Based on your suggestion, I simplified the description as: "A set of element IDs that are pre-selected as prototypes. Remaining prototypes are then recruited with the prototype selection algorithm. Warning: It will most likely violate the global objective function."

File: genomesubsampler/prototypeSelection.py (hunk 2 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Sure! Comment added.

File: genomesubsampler/prototypeSelection.py (hunk 3 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Thanks! Changed.

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: I think in general, Python tuple performs better than Python list due to its immutable nature.

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py:185]: I think that without including `seedset`, the result of `distance_sum` will be different, and doesn't guarantee picking the same next element. This is a problem we discussed a long time ago.

File: genomesubsampler/prototypeSelection.py (hunk 8 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Good idea! You may add it based on your benchmark scores.

File: genomesubsampler/prototypeSelection.py (hunk 7 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: You idea looks good, but `res_set = list(seedset)` doesn't work the same as the original implementation.

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: It won't matter, but I think the user would like to have more options.

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: you are probably right in general, but here I have the feeling that we only confuse the reader and the performance impact is so little

File: genomesubsampler/prototypeSelection.py (hunk 8 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: I don't understand. I am speaking about adding the sting ', seedset: List[str]' to the signature in the docstring.

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py:185]: You are right, but in a different context. Here, we iterate through all exponential many final sets of prototypes and pick the optimal one. Therefore, it won't change the choice on the optimal set if we only compute distance_sum(s), i.e. on a part of the set, because the preselected elements of seedset are guaranteed to be in the result. Thus, it really is only a waste of compute.
This is different from situations where we iteratively construct the prototype set element by element.

File: genomesubsampler/prototypeSelection.py (hunk 7 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: can you be more specific? Does it fail? Did it select different prototypes?

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py]: yeah, but you already restricted the options for the user by running you validation on the input parameters (which I think is a good idea).

File: genomesubsampler/prototypeSelection.py [hunk 12]
```diff
@@ -62,11 +109,10 @@ def distance_sum(elements, dm):
     function signature with type annotation for future use with python >= 3.5
     def distance_sum(elements: Sequence[str], dm: DistanceMatrix) -> float:
     '''
-
     return np.tril(dm.filter(elements).data).sum()
 
 
-def prototype_selection_exhaustive(dm, num_prototypes,
+def prototype_selection_exhaustive(dm, num_prototypes, seedset=[],
```
Comment (wasade) [inline genomesubsampler/prototypeSelection.py]: Please use `seedset=None` and test whether it is `None` in the body of the function. It is a dangerous pattern to set a default object like this, and can lead to unexpected behavior:

```python
>>> def foo(bar=[]):
...     bar.append('a')
...     print(bar)
...
>>> foo()
['a']
>>> foo()
['a', 'a']
>>> foo()
['a', 'a', 'a']
>>> foo()
['a', 'a', 'a', 'a']
```

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (wasade) [reply genomesubsampler/prototypeSelection.py]: The difference is in the microseconds for large structures. Microoptimizations should only be performed if supported by profiling data suggesting that this is a runtime bottleneck.

```python
In [222]: a = tuple(range(10000))

In [223]: b = list(range(10000))

In [224]: %timeit a + a
10000 loops, best of 3: 64.6 µs per loop

In [225]: %timeit b + list(a)
10000 loops, best of 3: 99.6 µs per loop
```

File: genomesubsampler/prototypeSelection.py (hunk 7 above)
Comment (wasade) [reply genomesubsampler/prototypeSelection.py]: `.index` is `O(N)`...

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (wasade) [inline genomesubsampler/prototypeSelection.py]: from what I can see in the code below, it does not seem to matter if `seedset` is or is not cast to `set`

Comment (wasade) [reply genomesubsampler/prototypeSelection.py]: I just deleted my last comment, I misread the code. Casting to `set` makes sense because of the lookup performed on line 244. If `seedset` is a `list` or `tuple` then it will break down to a linear search. On the assumption that `seedset` may realistically contain more than like 10 entries, it is probably better to just cast to `set` to take advantage of `O(1)` membership testing

**Comment by qiyunzhu** [comment]
Hello @sjanssen2 and @wasade Thank you for your review of my code! Special thanks to @wasade , your demo codes are very helpful. I would've never learn these tricks without you telling me. I was distracted by other projects for a while. Now I am getting back to improve this code.

File: genomesubsampler/prototypeSelection.py (hunk 4 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: @wasade thanks for doing this!

@sjanssen2 Excuse me for my unclear response. The reason I converted `seedset` to `tuple` is not only for performance, but also to make sure the subsequent lines, `d = distance_sum(s + seedset, dm)` and `return list(seedset + max_set)` will work without type incongruence error.

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py:185]: @sjanssen2 Well, I tested the case of `d = distance_sum(s, dm)` and, the resulting prototypes are different...

File: genomesubsampler/prototypeSelection.py (hunk 8 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Got it! Changed.

File: genomesubsampler/prototypeSelection.py (hunk 7 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: @sjanssen2 The difference is because `res_set` is NOT a list of `id`, but a list of `idx`.

File: genomesubsampler/prototypeSelection.py (hunk 6 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: @wasade I didn't see your previous comment. But this comment is very helpful to me!

File: genomesubsampler/prototypeSelection.py (hunk 12 above)
Comment (qiyunzhu) [reply genomesubsampler/prototypeSelection.py]: Thanks for explaining this! This is indeed something I can't learn from Google (since I didn't run into error I won't Google). I have changed all the instances accordingly.

File: genomesubsampler/prototypeSelection.py (hunk 5 above)
Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py:185]: hm. Interesting. Can you give me an example?

Comment (sjanssen2) [reply genomesubsampler/prototypeSelection.py:185]: Oh shit, you are so right and I overlooked something terrible. Sorry for my stupidity :-/

**Review by sjanssen2 — APPROVED** [review]
